staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go

```diff
@@ -164,6 +154,10 @@ func (a *QuotaAdmission) Validate(ctx context.Context, attr admission.Attributes
 	if attr.GetNamespace() == "" || isNamespaceCreation(attr) {
 		return nil
 	}
+	// we need to wait for our caches to warm
+	if !a.WaitForReady() {
```

Comment:
doing this here, before calling Evaluate(), means that in unsynced cases, we'll reject resources we would otherwise have ignored:

https://github.com/kubernetes/kubernetes/blob/d36737c817303e525e00e5ada24a7a79c05ebb53/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go#L601-L606

or skipped:

https://github.com/kubernetes/kubernetes/blob/d36737c817303e525e00e5ada24a7a79c05ebb53/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go#L617-L621

I think the WaitForReady call should happen *after* both of those, so we only block on sync for things quota is actually interested in. That will mean plumbing the wait function down into the evaluator to invoke.

Reply:
I moved it to the evaluator, but I am not sure if this is the best place. `init` can't be used as it doesn't let for retrying errors. Moving to `quotaAccessor` may work, but it would likely need a custom implementation of `waitForReady`:
```go
func (a *quotaAccessor) waitForReady() error {
  if a.hasSynced() {
    return nil
  }
  if err := wait.PollUntilContext(context.Background, 100 * time.Millisecond, time.Second, func(_ context.Context) { return a.hasSynced(), nil }); err != nil {
    return fmt.Errorf("not ready")
  }
```

I also hope that compiler manages to optimize handler's `WaitForReady` as otherwise it creates a timer only to be GCed right after (in cases when `readyFunc` returns always `true`, i.e. post sync)